src/core/qgsvectorlayerutils.cpp

```diff
@@ -380,8 +380,16 @@ QgsFeature QgsVectorLayerUtils::createFeature( const QgsVectorLayer *layer, cons
     bool checkUnique = true;
 
     // in order of priority:
+    // 1. passed attribute value
+    if ( attributes.contains( idx )
+         && !layer->primaryKeyAttributes().contains( idx )
```

Comment:
as said, I would completely remove any reference to primaryKey here, since they should be tagged as unique and in case of compound keys need special treatment anyway.

Reply:
Done!